lib/web_ui/lib/src/ui/painting.dart

```diff
@@ -1982,3 +2006,23 @@ class Shadow {
   @override
   String toString() => 'TextShadow($color, $offset, $blurRadius)';
 }
+
+/// Determines if lists [a] and [b] are deep equivalent.
+///
+/// Returns true if the lists are both null, or if they are both non-null, have
+/// the same length, and contain the same elements in the same order. Returns
+/// false otherwise.
+bool _listEquals<T>(List<T> a, List<T> b) {
```

Comment:
Since this is originally defined in `lib/ui/text.dart`, I wonder if should also put this inside `lib/web_ui/lib/src/ui/text.dart`.

Reply:
It looks like it lives in lib/web_ui/lib/src/engine/text/paragraph.dart on the web_ui side of things. I'll look into reusing that version.

Reply:
It is inside engine:paragraph.dart in the web_ui side. I don't think engine and ui can share an internal function in the same way that non-web painting.dart and text.dart can (both are in ui).

Is there a way to do that?

Reply:
It looks like painting.dart has an abstract ColorFilter class that defers to engine.EngineColorFilter for its work. The same pattern should probably be followed for ImageFilter, but it is currently all stubbed out in the web_ui files so I'll have to untangle that...

Reply:
This will no longer be needed after I update to the latest upstream master - the web team implemented an actual EngineImageFilter object and I'll have to undo all of my changes here and re-create them on the new class.